volar_virtual_file_hierarchy.py

```diff
@@ -0,0 +1,235 @@
+        return Promise.resolve(element['embeddedFiles'])
+
+    def get_tree_item(self, element: VirtualFile) -> TreeItem:
+        command_url = sublime.command_url('lsp_volar_open_virtual_file', {
+            'uri': self.source_file_uri,
+            'file_name': element['fileName'],
+        })
+        path = basename(element['fileName'])
+        description = '(kind: {}, version: {})'.format(element['kind'], element['version'])
+        return TreeItem(
+            path,
+            description=description,
+            command_url=command_url
+        )
+
+
+class LspVolarShowVirtualFilesCommand(LspTextCommand):
+    session_name = 'LSP-volar'
+
+    def run(self, edit: sublime.Edit) -> None:
+        sublime.set_timeout_async(self.run_async)
+
+    def run_async(self) -> None:
+        session = self.session_by_name()
+        if not session:
+            return
+        request = Request(GetVirtualFilesRequest.Type, text_document_identifier(self.view), progress=True)
+        session.send_request_task(request) \
+            .then(lambda virtual_file: self._on_get_virtual_files_async(weakref.ref(session), virtual_file))
+
+    def _on_get_virtual_files_async(
+        self, weaksession: 'weakref.ref[Session]', virtual_file: GetVirtualFilesRequest.ResponseType
+    ) -> None:
+        if not virtual_file:
+            sublime.status_message('No virtual file found')
+            return
+        window = self.view.window()
+        if not window:
+            return
+        session = weaksession()
+        if not session:
+            return
+        sheet_name = 'Virtual Files (LSP-volar)'
+        source_file_uri = uri_from_view(self.view)
+        elements = [virtual_file]
+        data_provider = VirtualFilesDataProvider(weaksession, source_file_uri, elements)
+        new_tree_view_sheet(window, sheet_name, data_provider, sheet_name)
+        window.run_command('lsp_volar_open_virtual_file', {
+            'uri': source_file_uri,
+            'file_name': virtual_file['fileName'],
+        })
+
+
+class LspVolarOpenVirtualFileCommand(LspWindowCommand):
+    session_name = 'LSP-volar'
+
+    def run(self, uri: str, file_name: str, event: Optional[dict] = None) -> None:
+        sublime.set_timeout_async(lambda: self.run_async(uri, file_name))
+
+    def run_async(self, uri: str, file_name: str) -> None:
+        session = self.session()
+        if not session:
+            return
+        request = Request(GetVirtualFileRequest.Type, {
+            'sourceFileUri': uri,
+            'virtualFileName': file_name,
+        })
+        session.send_request_task(request).then(lambda result: self._on_files_contents_async(uri, file_name, result))
+
+    def _on_files_contents_async(self, uri: URI, file_name: str, result: GetVirtualFileRequest.ResponseType) -> None:
+        flags = sublime.ADD_TO_SELECTION | sublime.SEMI_TRANSIENT | sublime.CLEAR_TO_RIGHT
+        # Force TS syntax for .js files
+        if file_name.endswith('.js'):
```

Comment:
can you write a comment on why you need force the ts syntax for js files?

Reply:
or leave the comment here on Github, 
just so someone can learn why this code needs to exist.

Reply:
I'm also trying to fix this on the Volar side at https://github.com/vuejs/language-tools/pull/3353 which explains the reasoning. But basically those files are typescript files and using `.js` extension triggers various errors from TypeScript.

But this doesn't really do much currently here because LSP doesn't initialize on those files anyway. The implement was a bit different when I've initially wrote that.